### PR TITLE
fix(iot-dev): Fix bug where AMQP layer would block a while on closing if the proton reactor thread crashed

### PR DESCRIPTION
**Pull request by timtay-microsoft**
#1379 

We had a basic assumption that onReactorFinal would always be called when proton-j's thread was finished, but if proton-j's thread encountered a runtime exception or some handler exception, the thread would exit prematurely and would never execute onReactorFinal. Because of that, we need to release some state latches when this happens so that the user's close thread doesn't block for 20 seconds as discussed in the issue above.

### CODE REVIEW

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsIotHubConnection.java [hunk 1]
```diff
@@ -1238,44 +1243,15 @@ private void executorServicesCleanup()
         }
     }
 
-    private static class ReactorRunner implements Callable<Object>
```
Comment (timtay-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsIotHubConnection.java:1241]: This change is a bit hard to follow, but basically I moved this class out to its own file so that this file didn't have other classes defined in it. I then moved all the logic from the IotHubReactor class into this class since IotHubReactor was an unnecessary layer to have here.

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/IotHubReactor.java [hunk 2]
```diff
@@ -1,39 +0,0 @@
...
-        this.reactor.start();
-
-        //noinspection StatementWithEmptyBody
-        while (this.reactor.process())
-        {
-            // The empty while loop is to ensure that reactor thread runs as long as it has messages to process
-        }
-        this.reactor.stop();
-        this.reactor.process();
-        this.reactor.free();
```
Comment (timtay-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/IotHubReactor.java:32]: This one function was the sole reason we had this class abstracting the Reactor. I moved this logic into the ReactorRunner class to cut down on the number of classes we have.

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/IotHubReactor.java [hunk 3]
```diff
@@ -1,39 +0,0 @@
...
-    }
-
-    public void free()
-    {
-        this.reactor.free();
```
Comment (timtay-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/IotHubReactor.java:37]: A perfect illustration of why this layer of abstraction was pretty pointless

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/ReactorRunner.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        {
+            TransportException transportException = new TransportException(e);
+
+            // unclassified exceptions are treated as retryable in ProtonJExceptionParser, so they should be treated
+            // the same way here. Exceptions caught here tend to be transient issues.
+            transportException.setRetryable(true);
+
+            // Notify the AMQP connection layer so it can tear down the reactor's resources that would usually
+            // get cleaned up in a graceful close of the reactor
+            this.reactorRunnerStateCallback.onReactorClosedUnexpectedly();
```
Comment (timtay-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/ReactorRunner.java:56]: This is the new stuff that actually fixes the bug. Just another state callback so that we can release the latches like the comment mentions.

**Review by jamdavi — APPROVED** [review]
